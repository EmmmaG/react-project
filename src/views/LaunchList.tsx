import * as React from 'react'
import { withStyles, WithStyles, Theme } from '@material-ui/core/styles'
import Typography from '@material-ui/core/Typography'

import { Launch, Launchpad } from './LaunchDataInterfaces'
import LaunchCard from '../components/LaunchCard'

const styles = (theme: Theme) => ({
	listContainer: {
		marginLeft: '35px',
		marginRight: '35px',
		backgroundColor: 'white',
		paddingTop: '20px',
	},
	titleContainer: {
		textAlign: 'center' as 'center',
	},
	title: {
		color: theme.palette.grey[600],
		fontSize: '12px',
		fontWeight: 'bold' as 'bold',
	},
})

export interface Props extends WithStyles<typeof styles> {
	launchesData: Launch[]
	launchpadsData: Launchpad[]
}

export class LaunchList extends React.Component<Props> {
	private renderLaunchList = () => {
		const { launchesData, launchpadsData } = this.props

		return launchesData.map(data =>
			<LaunchCard
				key={`${data.rocket.rocket_name} - ${data.payloads[0].payload_id}`}
				launchData={data}
				launchpadsData={launchpadsData}
			/>,
		)
	}

	public render() {
		const { classes, launchesData } = this.props

		return (
			<div className={classes.listContainer}>
				<div className={classes.titleContainer}>
					<Typography className={classes.title}>{`Showing ${launchesData.length} Missions`}</Typography>
				</div>
				{this.renderLaunchList()}
			</div>
		)
	}
}

export default withStyles(styles)(LaunchList)
