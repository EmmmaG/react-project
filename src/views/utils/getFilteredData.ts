import { Launch } from '../LaunchDataInterfaces'
import { FilterData } from '../LaunchMission'

export default function getFilteredData(launchesData: Launch[], filterData: FilterData) {

}
