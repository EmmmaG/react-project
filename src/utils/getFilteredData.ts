import { Launch } from '../views/LaunchDataInterfaces'
import { FilterData } from '../views/LaunchMission'

/**
 * to filter launches data with given filter options
 * @param launchesData data to be filtered
 * @param filterData data from selected filter
 */
export default function getFilteredData(launchesData: Launch[], filterData: FilterData) {
	let filteredData = launchesData
	// filter by keywords
	if (filterData.keywords) {
		const keywords = `${filterData.keywords}`.replace(/\s/g, '').toLowerCase()
		const filteredByKeywords = launchesData.filter(data =>
			`${data.flight_number}` === keywords
			|| `${data.rocket.rocket_name}`.toLowerCase().includes(keywords)
			|| `${data.payloads[0].payload_id}`.toLowerCase().includes(keywords),
		)

		filteredData = filteredByKeywords
	}

	// filter by launchPad
	if (filterData.launchPad !== 'any') {
		const filteredByLaunchPad = filteredData.filter(data => data.launch_site.site_id === filterData.launchPad)

		filteredData = filteredByLaunchPad
	}

	// filter by Min Year
	if (filterData.minYear !== 'any') {
		const filteredByMinYear = filteredData.filter(data => {
			const year = new Date(data.launch_date_local).getFullYear()

			return year >= parseInt(filterData.minYear, 10)
		})

		filteredData = filteredByMinYear
	}

	// filter by Max Year
	if (filterData.maxYear !== 'any') {
		const filteredByMaxYear = filteredData.filter(data => {
			const year = new Date(data.launch_date_local).getFullYear()

			return year <= parseInt(filterData.maxYear, 10)
		})

		filteredData = filteredByMaxYear
	}

	return filteredData
}
