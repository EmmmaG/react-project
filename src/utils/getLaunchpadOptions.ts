import { Launchpad } from '../views/LaunchDataInterfaces'
import { Props as DropdownSelectProps } from '../components/DropdownSelect'

/**
 * To grab launchPadOptions for filter from launchpadData fetched from server
 * @param launchpadData launchpadData get from server
 */
export default function getLaunchPadOptions(launchpadData: Launchpad[]): DropdownSelectProps['options'] {
	const launchPadOptions = launchpadData.map(data => ({
		value: data.id,
		text: data.full_name,
	}))

	launchPadOptions.unshift({
		value: 'any',
		text: 'Any',
	})

	return launchPadOptions
}
