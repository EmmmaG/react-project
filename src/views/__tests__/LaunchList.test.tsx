import * as React from 'react'
import { shallow } from 'enzyme'
import idObj from 'identity-obj-proxy'

import { launchpadData } from '../../utils/__tests__/getLaunchpadOptions.test'
import { launchData } from '../../utils/__tests__/getFilteredData.test'

import LaunchList, { Props } from '../LaunchList'

const props: Props = {
	classes: idObj,
	launchesData: launchData,
	launchpadsData: launchpadData,
}

describe('LaunchList', () => {
	it('renders correctly', () => {
		const wrapper = shallow(<LaunchList {...props} />)
		expect(wrapper).toMatchSnapshot()
	})
})
