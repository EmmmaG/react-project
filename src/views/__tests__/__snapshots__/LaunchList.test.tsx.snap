// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LaunchList renders correctly 1`] = `
<LaunchList
  classes={
    Object {
      "listContainer": "LaunchList-listContainer-1",
      "title": "LaunchList-title-3",
      "titleContainer": "LaunchList-titleContainer-2",
    }
  }
  launchesData={
    Array [
      Object {
        "flight_number": 1,
        "launch_date_local": "2006-03-25T10:30:00+12:00",
        "launch_site": Object {
          "site_id": "kwajalein_atoll",
          "site_name": "Kwajalein Atoll",
        },
        "payloads": Array [
          Object {
            "payload_id": "FalconSAT-2",
          },
        ],
        "rocket": Object {
          "rocket_id": "falcon1",
          "rocket_name": "Falcon 1",
          "rocket_type": "Merlin A",
        },
      },
      Object {
        "flight_number": 2,
        "launch_date_local": "2007-03-25T10:30:00+12:00",
        "launch_site": Object {
          "site_id": "ccafs_slc_40---",
          "site_name": "Kwajalein Atoll",
        },
        "payloads": Array [
          Object {
            "payload_id": "FalconSAT-3",
          },
        ],
        "rocket": Object {
          "rocket_id": "falcon1",
          "rocket_name": "Falcon 2",
          "rocket_type": "Merlin A",
        },
      },
      Object {
        "flight_number": 3,
        "launch_date_local": "2008-03-25T10:30:00+12:00",
        "launch_site": Object {
          "site_id": "ccafs_slc_40---",
          "site_name": "Kwajalein Atoll",
        },
        "payloads": Array [
          Object {
            "payload_id": "FalconSAT-33",
          },
        ],
        "rocket": Object {
          "rocket_id": "falcon1",
          "rocket_name": "Falcon 3",
          "rocket_type": "Merlin A",
        },
      },
      Object {
        "flight_number": 4,
        "launch_date_local": "2009-03-25T10:30:00+12:00",
        "launch_site": Object {
          "site_id": "kwajalein_atoll",
          "site_name": "Kwajalein Atoll",
        },
        "payloads": Array [
          Object {
            "payload_id": "FalconSAT-44",
          },
        ],
        "rocket": Object {
          "rocket_id": "falcon1",
          "rocket_name": "Falcon 4",
          "rocket_type": "Merlin A",
        },
      },
      Object {
        "flight_number": 5,
        "launch_date_local": "2010-03-25T10:30:00+12:00",
        "launch_site": Object {
          "site_id": "kwajalein_atoll",
          "site_name": "Kwajalein Atoll",
        },
        "payloads": Array [
          Object {
            "payload_id": "FalconSAT-55",
          },
        ],
        "rocket": Object {
          "rocket_id": "falcon1",
          "rocket_name": "Falcon 5",
          "rocket_type": "Merlin A",
        },
      },
      Object {
        "flight_number": 66,
        "launch_date_local": "2011-03-25T10:30:00+12:00",
        "launch_site": Object {
          "site_id": "kwajalein_atoll",
          "site_name": "Kwajalein Atoll",
        },
        "payloads": Array [
          Object {
            "payload_id": "FalconSAT-77",
          },
        ],
        "rocket": Object {
          "rocket_id": "falcon6",
          "rocket_name": "Falcon 6",
          "rocket_type": "Merlin A",
        },
      },
    ]
  }
  launchpadsData={
    Array [
      Object {
        "full_name": "Kwajalein Atoll",
        "id": "kwajalein_atoll",
      },
      Object {
        "full_name": "Cape Canaveral Air Force Station Space Launch Complex 40",
        "id": "ccafs_slc_40",
      },
    ]
  }
/>
`;
